### PR TITLE
Use the new re-entrant do-nothing script

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

Just a draft so we can trial it in the upcoming release.

https://github.com/SciTools-incubator/nothing

- [x] need to set appropriate permissions for the new repo

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

---
Add any of the below labels to trigger actions on this PR:

- https://github.com/SciTools/iris/labels/benchmark_this

### CODE REVIEW

File: tools/release_do_nothing.py [hunk 1]
```diff
@@ -10,879 +10,776 @@
...
-        whats_new_rsts,
-    )
-
-    _break_print("RELEASE COMPLETE. Congratulations! 🎉")
+            "Update the release page in GitHub discussions, with the above "
+            "links "
+            "and anything else appropriate.\n"
+            "https://github.com/SciTools/iris/discussions"
+        )
+        self.wait_for_done(message)
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply tools/release_do_nothing.py:755]: 76beb65a4

File: tools/release_do_nothing.py [hunk 2]
```diff
@@ -10,879 +10,776 @@
...
+
+    def twitter_announce(self):
+        message = (
+            "Announce the release via https://twitter.com/scitools_iris, "
+            "and any "
+            "other appropriate message boards (e.g. Yammer).\n"
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply tools/release_do_nothing.py]: a010bd3d9

File: tools/release_do_nothing.py [hunk 3]
```diff
@@ -10,879 +10,776 @@
...
+                '-m "Merging release branch into main";'
+            )
+            self.wait_for_done(message)
+
+            message = (
+                "Recreate the following files, which are present in "
+                "``main``, but "
+                f"are currently deleted from {working_branch}:\n"
+                f"{self.whats_news.latest.absolute()}\n"
+                f"{self.whats_news.template.absolute()}\n"
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 0c9681058

File: tools/release_do_nothing.py [hunk 4]
```diff
@@ -10,879 +10,859 @@
...
-    message = (
-        f"Follow the Pull Request process to get {release_strings.tag} branch "
-        f"merged into upstream/{upstream_branch} .\n"
-        "Specific conda-forge guidance will be automatically given once the "
-        "PR is created."
-    )
-    _wait_for_done(message)
+        def validate(sha256_string: str) -> str:
+            try:
+                sha_len = int(sha256_string, 16).bit_length() + 1
```
Comment (trexfeathers) [inline tools/release_do_nothing.py:517]: This doesn't actually work - the `bit_length()` of the string can be anywhere from `253` to `256`. The correct two things to check are:

- That the string can be validly parsed as a 16-bit number (`int(sha256_string, 16)`)
- That the string is 64 characters long

File: tools/release_do_nothing.py [hunk 5]
```diff
@@ -10,879 +10,859 @@
...
-        f"https://anaconda.org/conda-forge/iris?version={release_strings.release}\n"
-    )
-    _wait_for_done(message)
+        message = (
+            "Confirm that pip install works as expected:\n"
+            "conda create -y -n tmp_iris pip cf-units;\n"
+            "conda activate tmp_iris;\n"
+            f"pip install scitools-iris=={self.strings.release};\n"
+            'python -c "import iris; print(iris.__version__)";\n'
+            "conda remove -n tmp_iris --all;\n"
```
Comment (HGWright) [inline tools/release_do_nothing.py:544]: This requires a `"conda deactivate"` to work

File: tools/release_do_nothing.py [hunk 6]
```diff
@@ -10,879 +10,859 @@
...
-        _wait_for_done(message)
+        self.wait_for_done(message)
 
         message = (
-            "Commit and push all the What's New changes.\n"
-            "git commit -am \"Restore latest What's New files.\";\n"
-            f"git push -u origin {working_branch};"
+            "After the PR is merged, wait for the CI to complete, after which "
+            "the new version of Iris will be on conda-forge's servers.\n"
+            "https://github.com/conda-forge/iris-feedstock/actions"
```
Comment (trexfeathers) [inline tools/release_do_nothing.py:679]:
```suggestion
            "https://dev.azure.com/conda-forge/feedstock-builds/_build?definitionId=464"
```

File: tools/release_do_nothing.py [hunk 7]
```diff
@@ -10,879 +10,859 @@
...
 
-    _mark_section(4)
-    is_first_in_series = check_first_in_series(
-        release_type,
-        release_strings,
-        is_release_candidate,
-    )
+        message = (
+            f"Comment on {discussion_url} to notify anyone watching that "
+            f"{self.strings.tag} has been released."
```
Comment (trexfeathers) [inline tools/release_do_nothing.py:753]:
```suggestion
            f"{self.git_tag} has been released."
```

File: tools/release_do_nothing.py [hunk 8]
```diff
@@ -10,879 +10,859 @@
...
-    _break_print("Branch merge-back ...")
+                message = (
+                    "Follow the latest conda-forge guidance for creating a new "
+                    "release candidate branch from the `main` branch:\n"
+                    "https://conda-forge.org/docs/maintainer/knowledge_base.html#pre-release-builds\n\n"
+                    "DEVIATION FROM GUIDANCE: config file(s) should point to "
+                    "the `rc_iris` label (this is not the name that "
+                    "conda-forge suggest).\n"
+                )
+                rc_branch = self.get_input(message, "Input the name of your new branch")
```
Comment (trexfeathers) [inline tools/release_do_nothing.py:625]: The changes are best made within the same PR that proposes the new release.

File: tools/release_do_nothing.py (hunk 4 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py]: bc7f6bca2

File: tools/release_do_nothing.py (hunk 5 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py:547]: 0f8721b47

File: tools/release_do_nothing.py (hunk 6 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 78d0255c5

File: tools/release_do_nothing.py (hunk 7 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 94fa686b5

File: tools/release_do_nothing.py (hunk 8 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 25812b0c5

**Comment by pp-mo** [comment]
Tested this in use + it was handy
But it would be nice if the "nothing" package allowed you to restart from latest stored. 
I created a couple of issues to suggest improvements in the common code package : https://github.com/SciTools-incubator/nothing/issues/1 https://github.com/SciTools-incubator/nothing/issues/2

**Comment by trexfeathers** [comment]
> Tested this in use + it was handy But it would be nice if the "nothing" package allowed you to restart from latest stored. I created a couple of issues to suggest improvements in the common code package : [SciTools-incubator/nothing#1](https://github.com/SciTools-incubator/nothing/issues/1) [SciTools-incubator/nothing#2](https://github.com/SciTools-incubator/nothing/issues/2)

Thanks @pp-mo these are great suggestions! I don't consider them a blocker to this PR, which I know @HGWright is looking at today.

File: tools/release_do_nothing.py [hunk 9]
```diff
@@ -10,879 +10,869 @@
...
 import typing
 
+try:
+    from nothing import Progress
+except ImportError:
+    install_message = (
+        "This script requires the `nothing` package to be installed:\n"
+        "pip install git+https://github.com/SciTools-incubator/nothing.git"
+    )
+    raise ImportError(message)
```
Comment (ESadek-MO) [inline tools/release_do_nothing.py:25]:
```suggestion
    raise ImportError(install_message)
```

File: tools/release_do_nothing.py [hunk 10]
```diff
@@ -10,879 +10,869 @@
...
+            self.wait_for_done(message)
 
+            url = "https://cfconventions.org/Data/cf-standard-names/current/src/cf-standard-name-table.xml"
+            file = Path(__file__).parents[1] / "etc" / "cf-standard-name-table.xml"
+            message = (
+                "Update the CF standard names table to the latest version:\n"
+                f'wget "{url}" -O {file};'
+                f"git add {file};"
+                "git commit -m 'Update CF standard names table.';"
+                f"git push -u origin {working_branch};"
```
Comment (ESadek-MO) [inline tools/release_do_nothing.py:187]:
```suggestion
                f'wget "{url}" -O {file};\n'
                f"git add {file};\n"
                "git commit -m 'Update CF standard names table.';\n"
                f"git push -u origin {working_branch};"
```

File: tools/release_do_nothing.py [hunk 11]
```diff
@@ -10,879 +10,869 @@
...
-        f"Follow the Pull Request process to get {release_strings.tag} branch "
-        f"merged into upstream/{upstream_branch} .\n"
-        "Specific conda-forge guidance will be automatically given once the "
-        "PR is created."
-    )
-    _wait_for_done(message)
+                message = (
+                    "Follow the latest conda-forge guidance for creating a new "
+                    "release candidate branch from the `main` branch:\n"
+                    "https://conda-forge.org/docs/maintainer/knowledge_base.html#pre-release-builds\n\n"
```
Comment (trexfeathers) [inline tools/release_do_nothing.py:623]: Need to include a note to check that conda-forge automation has updated the feedstock `README` to reference the release candidate channel - this is considered evidence that things have worked correctly.

File: tools/release_do_nothing.py [hunk 12]
```diff
@@ -10,879 +10,869 @@
...
+            )
+            self.wait_for_done(message)
 
-    if first_in_series:
-        # TODO: automate
+        if self.is_release_candidate:
+            channel_command = " -c conda-forge/label/rc_iris "
+        else:
+            channel_command = " -c conda-forge "
+        channel_command += "--override-channels "
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply tools/release_do_nothing.py:707]:
```suggestion
```

File: tools/release_do_nothing.py [hunk 13]
```diff
@@ -10,879 +10,868 @@
...
+            )
+        self.wait_for_done(message)
+
+        def validate(sha256_string: str) -> str:
+            valid = True
+            try:
+                _ = int(sha256_string, 16)
+            except ValueError:
+                valid = False
+            valid = valid & len(sha256_string) == 64
```
Comment (ESadek-MO) [inline tools/release_do_nothing.py:521]:
```suggestion
            valid = valid and len(sha256_string) == 64
```
I'm still figuring out what `&` does exactly, I think it does bitwise comparisons, so it's comparing `1` (`TRUE`) to `1000000` (`64`).

**Review by ESadek-MO — CHANGES_REQUESTED** [review]
I think you're pretty much there. There are a few comments sitting on this PR, and I've added one of my own. I think once they're all done we merge this is, and if there's further issues we fix those when they arise.

File: tools/release_do_nothing.py [hunk 14]
```diff
@@ -10,879 +10,869 @@
...
+            f"git push -u origin {working_branch};"
         )
-        _wait_for_done(message)
+        self.wait_for_done(message)
 
         message = (
-            f"Review {rsts.release.name} to ensure it is a good reflection of "
-            f"what is new in {release_strings.series}."
+            f"Follow the Pull Request process to get {working_branch} "
+            f"merged into upstream/{self.strings.branch} .\n"
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 4cf07a99bedfa8ab0ffc7b79ee1df17b48db9f3d

File: tools/release_do_nothing.py (hunk 11 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py:689]: 8577b7914d251593bde0473e4d09ce493393120a

File: tools/release_do_nothing.py (hunk 13 above)
Comment (trexfeathers) [reply tools/release_do_nothing.py]: 3e6fd1f99f6e5188711ef28e9f2a7510258ec374

I have tested this

**Review by ESadek-MO — APPROVED** [review]

**Comment by trexfeathers** [comment]
Thanks @ESadek-MO!

**Pull request merged**